<a href="https://colab.research.google.com/github/ddbbh9/funk-a-dunk/blob/main/MP_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install mmh3
import mmh3

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
!pip install BitVector
import BitVector as bv

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 161.7/161.7 KB 3.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for BitVector: filename=BitVector-3.5.0-py3-none-any.whl size=43507 sha256=7f20a79e3f33bec44f95184391b0fc9e9d334f53c284b67b017677546cb657ee
  Stored in directory: /root/.cache/pip/wheels/44/da/46/940839fb37eed5ad403ab59d4f40e35c171aed030bc729df4d
Successfully built BitVector


In [ ]:
import matplotlib.pyplot as plt

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd

In [ ]:
def seededHash(value, seed=None):
    if seed:
        return mmh3.hash(value, seed)
    return mmh3.hash(value)

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/MP_2/IMDB top 1000.csv')
df.head()

Unnamed: 0                                              Title Certificate  \
0           0                 1. The Shawshank Redemption (1994)           R   
1           1                            2. The Godfather (1972)           R   
2           2                          3. The Dark Knight (2008)       PG-13   
3           3                   4. The Godfather: Part II (1974)           R   
4           4  5. The Lord of the Rings: The Return of the Ki...       PG-13   

  Duration                     Genre  Rate  Metascore  \
0  142 min                     Drama   9.3       80.0   
1  175 min              Crime, Drama   9.2      100.0   
2  152 min      Action, Crime, Drama   9.0       84.0   
3  202 min              Crime, Drama   9.0       90.0   
4  201 min  Action, Adventure, Drama   8.9       94.0   

                                         Description  \
0  Two imprisoned men bond over a number of years...   
1  The aging patriarch of an organized crime dyna...   
2  When the menace known as the Joker wreaks havo...   
3  The early life and career of Vito Corleone in ...   
4  Gandalf and Aragorn lead the World of Men agai...   

                                                Cast  \
0  Director: Frank Darabont | Stars: Tim Robbins,...   
1  Director: Francis Ford Coppola | Stars: Marlon...   
2  Director: Christopher Nolan | Stars: Christian...   
3  Director: Francis Ford Coppola | Stars: Al Pac...   
4  Director: Peter Jackson | Stars: Elijah Wood, ...   

                                 Info  
0   Votes: 2,295,987 | Gross: $28.34M  
1  Votes: 1,584,782 | Gross: $134.97M  
2  Votes: 2,260,649 | Gross: $534.86M  
3   Votes: 1,107,253 | Gross: $57.30M  
4  Votes: 1,614,369 | Gross: $377.85M

In [ ]:
df["Genre"] = df["Genre"].str.split(",") 

In [ ]:
df = df.explode("Genre") 
df.reset_index(inplace = True)
df

index  Unnamed: 0                               Title Certificate  \
0         0           0  1. The Shawshank Redemption (1994)           R   
1         1           1             2. The Godfather (1972)           R   
2         1           1             2. The Godfather (1972)           R   
3         2           2           3. The Dark Knight (2008)       PG-13   
4         2           2           3. The Dark Knight (2008)       PG-13   
...     ...         ...                                 ...         ...   
2503    997         997                     400. JFK (1991)           R   
2504    997         997                     400. JFK (1991)           R   
2505    998         998       301. Nights of Cabiria (1957)   Not Rated   
2506    999         999         302. Throne of Blood (1957)   Not Rated   
2507    999         999         302. Throne of Blood (1957)   Not Rated   

     Duration      Genre  Rate  Metascore  \
0     142 min      Drama   9.3       80.0   
1     175 min      Crime   9.2      100.0   
2     175 min      Drama   9.2      100.0   
3     152 min     Action   9.0       84.0   
4     152 min      Crime   9.0       84.0   
...       ...        ...   ...        ...   
2503  189 min    History   8.0       72.0   
2504  189 min   Thriller   8.0       72.0   
2505  110 min      Drama   8.1        NaN   
2506  110 min      Drama   8.1        NaN   
2507  110 min    History   8.1        NaN   

                                            Description  \
0     Two imprisoned men bond over a number of years...   
1     The aging patriarch of an organized crime dyna...   
2     The aging patriarch of an organized crime dyna...   
3     When the menace known as the Joker wreaks havo...   
4     When the menace known as the Joker wreaks havo...   
...                                                 ...   
2503  New Orleans District Attorney Jim Garrison dis...   
2504  New Orleans District Attorney Jim Garrison dis...   
2505  A waifish prostitute wanders the streets of Ro...   
2506  A war-hardened general, egged on by his ambiti...   
2507  A war-hardened general, egged on by his ambiti...   

                                                   Cast  \
0     Director: Frank Darabont | Stars: Tim Robbins,...   
1     Director: Francis Ford Coppola | Stars: Marlon...   
2     Director: Francis Ford Coppola | Stars: Marlon...   
3     Director: Christopher Nolan | Stars: Christian...   
4     Director: Christopher Nolan | Stars: Christian...   
...                                                 ...   
2503  Director: Oliver Stone | Stars: Kevin Costner,...   
2504  Director: Oliver Stone | Stars: Kevin Costner,...   
2505  Director: Federico Fellini | Stars: Giulietta ...   
2506  Director: Akira Kurosawa | Stars: Toshirô Mifu...   
2507  Director: Akira Kurosawa | Stars: Toshirô Mifu...   

                                    Info  
0      Votes: 2,295,987 | Gross: $28.34M  
1     Votes: 1,584,782 | Gross: $134.97M  
2     Votes: 1,584,782 | Gross: $134.97M  
3     Votes: 2,260,649 | Gross: $534.86M  
4     Votes: 2,260,649 | Gross: $534.86M  
...                                  ...  
2503     Votes: 139,634 | Gross: $70.41M  
2504     Votes: 139,634 | Gross: $70.41M  
2505       Votes: 42,160 | Gross: $0.75M  
2506                       Votes: 45,579  
2507                       Votes: 45,579  

[2508 rows x 11 columns]

In [ ]:
df['Genre'] = df['Genre'].str.strip()

In [ ]:
df['Genre'].value_counts()

Drama        728
Adventure    211
Comedy       208
Action       207
Crime        176
Thriller     165
Romance      147
Biography    129
Mystery      103
Animation    100
History       66
Sci-Fi        63
Fantasy       53
War           37
Family        30
Music         28
Film-Noir     17
Horror        13
Musical       12
Sport          8
Western        7
Name: Genre, dtype: int64

In [ ]:
set_drama = set(df.loc[(df['Genre'] == 'Drama')]['Description'])
set_comedy = set(df.loc[df['Genre'] == 'Comedy']['Description'])
set_adv = set(df.loc[df['Genre'] == 'Adventure']['Description'])
set_action = set(df.loc[df['Genre'] == 'Action']['Description'])
set_thr = set(df.loc[df['Genre'] == 'Thriller']['Description'])
set_rom = set(df.loc[df['Genre'] == 'Romance']['Description'])
set_bio = set(df.loc[df['Genre'] == 'Biography']['Description'])
set_mys = set(df.loc[df['Genre'] == 'Mystery']['Description'])
set_ani = set(df.loc[df['Genre'] == 'Animation']['Description'])
set_his = set(df.loc[df['Genre'] == 'History']['Description'])
set_sci =  set(df.loc[df['Genre'] == 'Sci-Fi']['Description'])
set_fan = set(df.loc[df['Genre'] == 'Fantasy']['Description'])
set_war  = set(df.loc[df['Genre'] == 'War']['Description'])
set_fam = set(df.loc[df['Genre'] == 'Family']['Description'])
set_ms = set(df.loc[df['Genre'] == 'Music']['Description'])
set_fn = set(df.loc[df['Genre'] == 'Comedy']['Description'])
set_hor = set(df.loc[df['Genre'] == 'Film-Noir']['Description'])
set_mus = set(df.loc[df['Genre'] == 'Musical']['Description'])
set_sp = set(df.loc[df['Genre'] == 'Sport']['Description'])
set_west = set(df.loc[df['Genre'] == 'Western']['Description'])

In [ ]:
set_list = [
    set_drama, 
    set_comedy, 
    set_adv, 
    set_action, 
    set_thr, 
    set_rom, 
    set_bio, 
    set_mys, 
    set_ani, 
    set_his, 
    set_sci, 
    set_fan,
    set_war, 
    set_fam,
    set_ms, 
    set_fn,
    set_hor,
    set_mus,
    set_sp,
    set_west
    ]

In [ ]:
# True Cardinality is 398
def kmin (k,s = None):
  hashed = []

  for d in df['Description']:
    hashed.append(seededHash(d,s))
    
  hashed.sort() 
  min = hashed[k]

  return min

def estimateN (k,min):
    if min != 0:
      N = (k/min) - 1

      return N


In [ ]:
m_vals = []
count = 0

estimates = []
#while count < 1000:

  
for i in range(100, 1100, 100):
  m_val = kmin(10,277)
  m_vals.append(m_val)


print(estimateN(100 ,m_vals[0]))



-1.0000000470219568


### Justification for M (the number of bits to give to the filter)

As we learned in class, there is an equation that can help us determine the optimal parameters to create the best possible filter. This equation is:

$k^* = ln 2$ * $(m/n)$

There are already two parameters given to us. The number of hashes (2) and the number of items we are hashing which is around 2,508 items. When plugging it into the equation we can figure out that:

2 = ln 2 * (m/2508)

m $\approx$ 7236.55 $\approx$ 7237 or 7237 bits


### Justification for K
 When calculating the Jaccard coefficients for the the set, all of them stopped growing after K = 201 (I checked this with other sets but in the interest of time I cannot demonstrate it with a graph.) Therefore any K's bigger than 201 would not improve the jaccard coefficient.


In [178]:
for j in range(1,1100, 100):
  a = minHash(set_drama,j, seed = 0)
  b = minHash(set_comedy, j, seed = 0)
  print(jaccard(a,b))
  print(j)

0.0
1
0.06626506024096386
101
0.0992063492063492
201
0.11572700296735905
301
0.11572700296735905
401
0.11572700296735905
501
0.11572700296735905
601
0.11572700296735905
701
0.11572700296735905
801
0.11572700296735905
901
0.11572700296735905
1001


In [ ]:
 def minHash(inSet, k, seed=None):
    hashed = []
    
    for val in inSet:
        hashed.append(seededHash(val,seed))
    h = hashed.sort()
    h_f = hashed[:k]
    
    return set(h_f)

### Minhash

In [182]:
mh1 = minHash(set_drama, 201, seed = 0)
mh2 = minHash(set_adv, 201, seed = 0)
mh3 = minHash(set_thr, 201, seed = 0)
mh4 = minHash(set_action, 201, seed = 0)
mh5 = minHash(set_comedy, 201, seed = 0)
l = [mh1, mh2, mh3, mh4, mh5]

In [183]:
j = []
for i in l:
  print('i:', i)
  for j in l:
    print('j:', j)
    print(jaccard(i,j))

i: {-332525566, -1995577852, -420240380, -1812995065, -513687032, -961597943, 319229968, -1901672431, -2123672045, -1752586733, -510713837, -1887161321, -1641628645, -1214758884, 4712987, 311036955, -1980956638, -1122827229, -1320118235, -1397903320, -1718089171, -1932550608, -890527180, 683351102, -459993025, 496170054, -1437271480, -1138452400, -916270511, -1976365486, 668830807, -1603570596, -605313444, -982240672, -1151486879, -1747104670, -583141276, 562126437, -265542550, -1428997011, -386891667, -393375121, -184863122, -204513165, -1522330506, -1913882500, -683014530, 14187135, -792574336, 541706366, -1262816118, 446689419, -1158368113, -126956913, -1580008302, -452134765, 396908691, 6361749, -763306346, -641146218, -506883434, -370528102, -2010277221, -1845255009, 266998944, 85954211, -254995796, -1861809489, -2047192911, -212659020, -246996807, -1030179654, -1754083653, 403696831, 756477124, 446019782, 44321479, -455330612, -238043441, -1575617327, -1570008366, -896359211, -16

In [ ]:
def jaccard(inSet1, inSet2):
    inter = inSet1.intersection(inSet2)
    union = inSet1.union(inSet2)
    jacc = len(inter)/len(union)

    return jacc

In [ ]:
def bloomFilter(inSet, m, hs1=0, hs2=10):
    
    final_bf = bv.BitVector(size = m)
    
    for val in inSet:
        h_1 = seededHash(val,hs1)
        final_bf[h_1 % m] = 1
        h_2 = seededHash(val,hs2)
        final_bf[h_2 % m] = 1
    
    return final_bf

In [ ]:
def bf_jaccard(bf1, bf2):
    inter = 0
    union = 0 
    for val in range(len(bf1)):
        if (bf1[val] == 1) and (bf2[val] == 1):
            inter += 1
        if (bf1[val] == 1) or (bf2[val] == 1):
            union += 1
    
    jacc = inter/union
    
    return jacc

In [ ]:
def bf_find(bf, val, hs1=0, hs2=10):
    
    ind_1 = seededHash(val, hs1) % len(bf)
    ind_2 = seededHash(val, hs2) % len(bf)
    
    if (bf[ind_1] == 1) and (bf[ind_2] == 1):
        return True
    else:
        return False